### PR TITLE
Add support for extras to pip-compile

### PR DESCRIPTION
**Pull request by paulkernfeld**
This addresses issue #154, "pip-compile does not handle extra requirements."

### CODE REVIEW

**Comment by mattrobenolt** [comment]
fwiw, I just commented out the check, and it appeared to work...

I was just about to submit a patch for:

```diff
diff --git a/piptools/resolver.py b/piptools/resolver.py
index 8892175..811edb2 100644
--- a/piptools/resolver.py
+++ b/piptools/resolver.py
@@ -98,10 +98,6 @@ class Resolver(object):
                 msg = ('pip-compile does not support URLs as packages, unless they are editable. '
                        'Perhaps add -e option?')
                 raise UnsupportedConstraint(msg, constraint)
-            elif constraint.extras:
-                msg = ('pip-compile does not yet support packages with extras. '
-                       'Support for this is in the works, though.')
-                raise UnsupportedConstraint(msg, constraint)
 
     def _group_constraints(self, constraints):
         """
```

then saw this. Not entirely sure why it worked for me without your patch.

**Comment by paulkernfeld** [comment]
Hey @mattrobenolt, it works in some situations without my patch, but it will fail in other cases. I think that it will fail in any of these cases:
1. If my project requires package A, and package A requires package B with with extras, pip-compile will not put the extras for package B into `requirements.txt`.
2. pip-compile doesn't know how to merge extras correctly. For example, it knows that it should merge `ipython>=2` and `ipython<3` into `ipython>=2,<3`, but it doesn't know that `ipython[notebook]` and `ipython[nbconvert]` should be merged into `ipython[nbconvert,notebook]`.
3. The pip-compile dependencies cache is keyed on only name and version (not on extras), so the cache doesn't know that `ipython[notebook]` and `ipython` have different dependencies. I actually just realized that my pull request does not fix this problem, so I need to update it.

**Comment by mattrobenolt** [comment]
Ah, I tested against a very simple case which worked fine. :)

File: tests/fixtures/fake-index.json [hunk 1]
```diff
@@ -29,6 +29,24 @@
       "Six"
     ]
   },
+  "gnureadline": {
+    "6.3.3": []
+  },
+  "ipython": {
+    "2.1.0": {
```
Comment (nvie) [inline tests/fixtures/fake-index.json]: I see you changed the data type under the version key to a dict here, which makes sense. To not further complicate the logic in the FakeRepository helper, let's just use this convention everywhere in this JSON file, so make all the other lists dicts (with only a `''` key), too.

**Comment by nvie** [comment]
Thanks, @paulkernfeld! Your analysis is spot-on, and I really look forward to properly support these. If you can address your point 3 (i.e. change the format of the dependency cache), then I'm happy to merge this in!

**Comment by paulkernfeld** [comment]
All right, @nvie, let me know what you think of this update! Sorry it's so long; I can split it into smaller commits if that's helpful.

**Comment by habnabit** [comment]
Found a bug in this PR, which I haven't written tests for but at least resolved: habnabit/pip-tools@973e62391656e20a26c724f9eca67f52caf26a43

**Comment by paulkernfeld** [comment]
@habnabit you're right, thanks for finding and fixing that! That is what I get for making changes without testing them :(

**Comment by paulkernfeld** [comment]
I have added some tests for the cache, and incorporated habnabit's fix. Right now I think Travis is broken, but the build passes on my computer.

**Comment by paulkernfeld** [comment]
Hey @nvie, is there anything I can do to make it easier to get this patch in?

File: tests/test_cache.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    # Create the file and write to it
+    cache_file = NamedTemporaryFile(mode="w", delete=False)
+    cache_file.write(to_write)
+    cache_file.close()
+
+    # Yield the path to the file
+    yield cache_file.name
+
+    # Delete the file on exit
+    remove(cache_file.name)
```
Comment (nvie) [inline tests/test_cache.py]: We should probably wrap this yield in a `try: ... finally:` block.

File: tests/test_cache.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    """
+    A good cache file.
+    """
+    with _read_cache_file_helper('{"__format__": 1, "dependencies": "success"}') as cache_file_name:
+        assert "success" == read_cache_file(cache_file_name)
+
+
+def test_reverse_dependencies(from_line):
+    # Since this is a test, make a temporary directory
+    tmp_dir_path = mkdtemp()
```
Comment (nvie) [inline tests/test_cache.py]: py.test offers a super nice way of creating temp dirs: you can simply have it inject the fixture `tmpdir` to this function, and you can use it by saying `tmp_dir_path = str(tmpdir)`. It'll be cleaned up automatically by pytest.

Comment (nvie) [inline tests/test_cache.py]: See https://pytest.org/latest/tmpdir.html

File: tests/test_resolver.py [hunk 4]
```diff
@@ -15,13 +15,57 @@
...
+             # Note that the extras should be sorted
+             'ipython[nbconvert,notebook]==2.1.0',
+             'pyzmq==2.1.12',
+             'jinja2==2.7.3',
+             'tornado==3.2.2',
+             'markupsafe==0.23',
+             'gnureadline==6.3.3',
+             'pygments==1.5',
+             'sphinx==0.3']
+         ),
```
Comment (nvie) [inline tests/test_resolver.py]: This looks great!

**Comment by nvie** [comment]
This is great work—thanks a lot, @paulkernfeld! Sorry I didn't find the time to review this a bit earlier. I had one case locally that behaved a bit strange. If I create a test file with the following contents, it still compiles all the extras, where I expected just the main `ipython` package:

    ipython#[notebook]

Note: there's no space between the 'n' and the '#'. I'd assume this would be considered a comment, but it wasn't. If you add a space, it acts how I would expect (the `#[notebook]` is considered a comment).

Otherwise, this looks fine, and I'm happy to merge it in.

**Comment by paulkernfeld** [comment]
Oh, very interesting point about the hash symbol. This is something that I know nothing about. Basically, I'm calling `pip.req.InstallRequirement.from_line`, and it ignores the hash symbol before the extras. We can definitely work around this, e.g. by only sending it the text before the hash symbol, but are you sure that's what we want to do?

I sort of think that this is the intended behavior, because when I run `pip install -r requirements.txt`, and `requirements.txt` has the one line, `ipython#[notebook]`, pip installs the notebook extra! Is this perhaps a bug in pip's parsing code? Or... a feature?

**Comment by paulkernfeld** [comment]
Oh boy. Actually, it looks like pip 6 considers `ipython#[notebook]` to mean `ipython`, whereas pip 7 considers it to mean `ipython[notebook]`.

File: tests/test_cache.py (hunk 3 above)
Comment (paulkernfeld) [inline tests/test_cache.py]: Thanks for the tip! I implemented this.

File: tests/test_cache.py (hunk 2 above)
Comment (paulkernfeld) [inline tests/test_cache.py]: You're right. Good point.

**Comment by nvie** [comment]
OK, I don't care in that case. If this is a bug in pip, it'll probably be fixed. If it's not, then we're just pip-compliant. No big deal.

Thanks for the work, I think it's awesome :)

**Comment by paulkernfeld** [comment]
Yay! Thanks @nvie!

**Pull request merged**